pkg/controller/direct/privilegedaccessmanager/entitlement_controller.go

```diff
@@ -0,0 +1,442 @@
+	}
+
+	a.actual = entitlementpb
+	return true, nil
+}
+
+func (a *Adapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	u := createOp.GetUnstructured()
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating PrivilegedAccessManagerEntitlement", "name", a.id.FullyQualifiedName())
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := PrivilegedAccessManagerEntitlementSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	req := &privilegedaccessmanagerpb.CreateEntitlementRequest{
+		Parent:        a.id.Parent.String(),
+		EntitlementId: a.id.Entitlement,
+		Entitlement:   resource,
+	}
+	op, err := a.gcpClient.CreateEntitlement(ctx, req)
+	if err != nil {
+		return fmt.Errorf("error creating PrivilegedAccessManagerEntitlement %s: %w", a.id.FullyQualifiedName(), err)
+	}
+	created, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("error waiting for PrivilegedAccessManagerEntitlement %s to be created: %w", a.id.FullyQualifiedName(), err)
+	}
+	log.V(2).Info("successfully created PrivilegedAccessManagerEntitlement", "name", a.id.FullyQualifiedName())
+
+	status := &krm.PrivilegedAccessManagerEntitlementStatus{}
+	observedState := PrivilegedAccessManagerEntitlementStatusObservedState_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ObservedState = observedState
+	status.ExternalRef = a.id.AsExternalRef()
+	return setStatus(u, status)
+}
+
+func (a *Adapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	u := updateOp.GetUnstructured()
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("updating PrivilegedAccessManagerEntitlement", "name", a.id.FullyQualifiedName())
+	mapCtx := &direct.MapContext{}
+
+	updateMask := &fieldmaskpb.FieldMask{}
+
+	if !reflect.DeepEqual(AdditionalNotificationTargets_FromProto(mapCtx, a.actual.AdditionalNotificationTargets), a.desired.Spec.AdditionalNotificationTargets) {
+		log.V(2).Info("'spec.additionalNotificationTargets' field is updated (-old +new)", cmp.Diff(AdditionalNotificationTargets_FromProto(mapCtx, a.actual.AdditionalNotificationTargets), a.desired.Spec.AdditionalNotificationTargets))
+		updateMask.Paths = append(updateMask.Paths, "additional_notification_targets")
+	}
+	if !reflect.DeepEqual(ApprovalWorkflow_FromProto(mapCtx, a.actual.ApprovalWorkflow), a.desired.Spec.ApprovalWorkflow) {
+		log.V(2).Info("'spec.approvalWorkflow' field is updated (-old +new)", cmp.Diff(ApprovalWorkflow_FromProto(mapCtx, a.actual.ApprovalWorkflow), a.desired.Spec.ApprovalWorkflow))
+		updateMask.Paths = append(updateMask.Paths, "approval_workflow")
+	}
+	if !reflect.DeepEqual(direct.SliceOfPointers_FromProto(mapCtx, a.actual.EligibleUsers, AccessControlEntry_FromProto), a.desired.Spec.EligibleUsers) {
```

Comment:
Did you try making EligibleUsers into []T, rather than []*T (I assume that is why you needed the `SliceOfPointers_...` helpers?)

Reply:
Refactored it in https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/2659.